zookeeper-specifications/zab-1.0/ZabWithFLE.tla

```diff
@@ -0,0 +1,1020 @@
+        /\ cepochRecv' = [cepochRecv EXCEPT ![i] = IF j \in cepochRecv[i] THEN cepochRecv[i] \ {j} ELSE cepochRecv[i]]
+        /\ ackeRecv'   = [ackeRecv   EXCEPT ![i] = IF j \in ackeRecv[i] THEN ackeRecv[i] \ {j} ELSE ackeRecv[i]]
+        /\ ackldRecv'  = [ackldRecv  EXCEPT ![i] = IF j \in ackldRecv[i] THEN ackldRecv[i] \ {j} ELSE ackldRecv[i]]
+        /\ ackIndex'   = [ackIndex   EXCEPT ![i][j] = 0]
+        /\ committedCounter' = [committedCounter EXCEPT ![i][j] = 0]
+        (* The action of follower j. *)
+        /\ FollowerShutdown(j)
+        (* Clean input buffer.*)
+        /\ Clean(i, j)
+        /\ UNCHANGED <<acceptedEpoch, history, commitIndex, currentCounter, sendCounter, committedIndex,
+                       tempVarsZ, cepochSent, synced, verifyVarsZ>>
+-----------------------------------------------------------------------------
+\* In phase f11, follower sends f.p to leader via FOLLOWERINFO(CEPOCH).
+FollowerSendFOLLOWERINFO(i) ==
+        /\ state[i]    = FOLLOWING
+        /\ zabState[i] = DISCOVERY
+        /\ leaderAddr[i] /= NullPoint
+        /\ \lnot cepochSent[i]
+        /\ Send(i, leaderAddr[i], [mtype  |-> FOLLOWERINFO,
+                                   mepoch |-> acceptedEpoch[i]])
+        /\ cepochSent' = [cepochSent EXCEPT ![i] = TRUE]
+        /\ UNCHANGED <<serverVarsZ, leaderVarsZ, electionVarsZ, tempVarsZ, leaderAddr, synced, verifyVarsZ, electionMsgs, idTable>>
+        
+(* In phase l11, leader waits for receiving FOLLOWERINFO from a quorum,
+   and then chooses a new epoch e' as its own epoch and broadcasts LEADERINFO. *)
+LeaderHandleFOLLOWERINFO(i, j) ==
+        /\ state[i] = LEADING
+        /\ msgs[j][i] /= << >>
+        /\ msgs[j][i][1].mtype = FOLLOWERINFO
+        /\ LET msg == msgs[j][i][1]
+           IN \/ /\ NullPoint \notin cepochRecv[i]   \* 1. has not broadcast LEADERINFO - modify tempMaxEpoch
+                 /\ LET newEpoch == Maximum({tempMaxEpoch[i], msg.mepoch})
+                    IN tempMaxEpoch' = [tempMaxEpoch EXCEPT ![i] = newEpoch]
+                 /\ Discard(j, i)
+              \/ /\ NullPoint \in cepochRecv[i]      \* 2. has broadcast LEADERINFO - no need to handle the msg, just send LEADERINFO to corresponding server
+                 /\ Reply(i, j, [mtype  |-> LEADERINFO,
+                                 mepoch |-> acceptedEpoch[i]])
+                 /\ UNCHANGED tempMaxEpoch
+        /\ cepochRecv' = [cepochRecv EXCEPT ![i] = IF j \in cepochRecv[i] THEN cepochRecv[i]
+                                                                          ELSE cepochRecv[i] \union {j}]
+        /\ UNCHANGED <<serverVarsZ, followerVarsZ, electionVarsZ, initialHistory, leadingVoteSet, learners, ackeRecv, ackldRecv, 
+                       forwarding, ackIndex, currentCounter, sendCounter, committedIndex, committedCounter, verifyVarsZ, electionMsgs, idTable>>
+
+LeaderDiscovery1(i) ==
```

Comment:
Action name is not descriptive, perhaps rename to something like LeaderBroadcastsLEADERINFO